core/node/core.go

```diff
@@ -0,0 +1,117 @@
+package node
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/go-ipfs/pin"
+	"github.com/ipfs/go-ipfs/repo"
+
+	"github.com/ipfs/go-bitswap"
+	"github.com/ipfs/go-bitswap/network"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipfs/go-ipfs-exchange-interface"
+	"github.com/ipfs/go-ipfs-exchange-offline"
+	"github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-mfs"
+	"github.com/ipfs/go-unixfs"
+	"github.com/libp2p/go-libp2p-host"
+	"github.com/libp2p/go-libp2p-routing"
+	"go.uber.org/fx"
+)
+
+func BlockServiceCtor(lc fx.Lifecycle, bs blockstore.Blockstore, rem exchange.Interface) blockservice.BlockService {
+	bsvc := blockservice.New(bs, rem)
+
+	lc.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			return bsvc.Close()
+		},
+	})
```

Comment:
So, this is going to be a bit error prone. If some start hook fails, we'll never shut this service down. I think the only _clean_ way to do this is to add a `Start` function.

We may want a helper for this:

```go
func CloseOnStop(lc fx.Lifecycle, c io.Closer) {
  lc.Apend(fx.Hook{
    OnStop: func(ctx context.Context) error {
      return c.Close()
    }
  })
}
```

Reply:
Looking at https://github.com/uber-go/fx/blob/master/app.go#L590, it seems like `OnStop` hooks are called if `OnStart` hooks fail (only the ones that have started - https://github.com/uber-go/fx/blob/master/internal/lifecycle/lifecycle.go#L60-L92)

Reply:
My worry is that we're actually _starting_ this service before `OnStart`. That is, we don't have separate construct/start steps.

Reply:
As far as I can tell, there isn't really anything actively running when we 'start' the block service. `blockservice.Close` only calls `Close` on exchange, which we also do in `OnlineExchangeCtor` few lines below.

We could wrap blockservice into an additional layer which returns an error on all calls until some `start()` method is called, which would expose any places using blockservice before it should be used.

Reply:
Meh. It's probably fine. I'm just wondering if we should move to a model where we don't actually _start_ systems till we call some `Start()` method. But we can handle that later.